os/win/dnd.cpp

```diff
@@ -0,0 +1,335 @@
+// LAF OS Library
+// Copyright (C) 2024  Igara Studio S.A.
+//
+// This file is released under the terms of the MIT license.
+// Read LICENSE.txt for more information.
+
+#include "os/window.h"
+#include "os/win/dnd.h"
+#include "os/system.h"
+
+#include "clip/clip.h"
+#include "clip/clip_win.h"
+
+#include <shlobj.h>
+
+DWORD as_dropeffect(const os::DropOperation op)
+{
+  DWORD effect = DROPEFFECT_NONE;
+  if (op & os::DropOperation::Copy)
+    effect |= DROPEFFECT_COPY;
+
+  if (op & os::DropOperation::Move)
+    effect |= DROPEFFECT_MOVE;
+
+  if (op & os::DropOperation::Link)
+    effect |= DROPEFFECT_LINK;
+
+  return effect;
+}
+
+os::DropOperation as_dropoperation(DWORD pdwEffect)
+{
+  int op = 0;
+  if (pdwEffect & DROPEFFECT_COPY)
+    op |= os::DropOperation::Copy;
+
+  if (pdwEffect & DROPEFFECT_MOVE)
+    op |= os::DropOperation::Move;
+
+  if (pdwEffect & DROPEFFECT_LINK)
+    op |= os::DropOperation::Link;
+
+  return static_cast<os::DropOperation>(op);
+}
+
+
+gfx::Point drag_position(HWND hwnd, POINTL& pt)
+{
+  ScreenToClient(hwnd, (LPPOINT) &pt);
+  return gfx::Point(pt.x, pt.y);
+}
+
+namespace os {
+
+// HGLOBAL Locking/Unlocking wrapper
+template <typename T>
+class GLock {
+public:
+  GLock() = delete;
+  GLock(const GLock&) = delete;
+  GLock(HGLOBAL hglobal) : m_hmem(hglobal) {
+    m_data = static_cast<T>(GlobalLock(m_hmem));
+  }
+
+  ~GLock() {
+    GlobalUnlock(m_hmem);
+  }
+
+  operator HGLOBAL() {
+    return m_hmem;
+  }
+
+  operator T () {
+    return m_data;
+  }
+
+  SIZE_T size() {
+    return GlobalSize(m_hmem);
+  }
+
+private:
+  HGLOBAL m_hmem;
+  T m_data;
+};
+
+base::paths DragDataProviderWin::getPaths()
+{
+  base::paths files;
+  FORMATETC fmt;
+  fmt.cfFormat = CF_HDROP;
+  fmt.ptd = nullptr;
+  fmt.dwAspect = DVASPECT_CONTENT;
+  fmt.lindex = -1;
+  fmt.tymed = TYMED::TYMED_HGLOBAL;
+  STGMEDIUM medium;
+  if (m_data->GetData(&fmt, &medium) == S_OK) {
+    {
+      GLock<HDROP> hdrop(medium.hGlobal);
+      if (static_cast<HDROP>(hdrop)) {
+        int count = DragQueryFile(hdrop, 0xFFFFFFFF, nullptr, 0);
+        for (int index = 0; index < count; ++index) {
+          int length = DragQueryFile(hdrop, index, nullptr, 0);
+          if (length > 0) {
+            std::vector<TCHAR> str(length + 1);
+            DragQueryFile(hdrop, index, &str[0], str.size());
+            files.push_back(base::to_utf8(&str[0]));
```

Comment:
We can add a comment about this `length+1`:
```suggestion
            // From Win32 docs: https://learn.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-dragqueryfilew
            // the DragQueryFile() does't include the null character in its return value.
            std::vector<TCHAR> str(length + 1);
            DragQueryFile(hdrop, index, str.data(), str.size());
            files.push_back(base::to_utf8(str.data()));
```
I have changed `&str[0]` to `str.data()` because clang-tidy prefer that (this file was't clang-tidy'd because it runs in the Linux port only).